File: app/src/main/java/org/stypox/dicio/settings/IOFragment.java
package org.stypox.dicio.settings;

import android.os.Bundle;

import org.stypox.dicio.R;
import org.stypox.dicio.input.VoskInputDevice;

import androidx.preference.PreferenceFragmentCompat;

public class IOFragment extends PreferenceFragmentCompat {
    @Override
    public void onCreatePreferences(final Bundle savedInstanceState, final String rootKey) {
        addPreferencesFromResource(R.xml.pref_io);

        findPreference(getString(R.string.pref_key_language))
                .setOnPreferenceChangeListener((preference, newValue) -> {
                    VoskInputDevice.deleteCurrentModel(requireContext());
                    if (getActivity() != null) {
                        getActivity().recreate();
                    }
                    return true;
                });
//TODO Discuss whether this is needed. At least for debugging commented
//        findPreference(getString(R.string.pref_key_input_method))
//                .setOnPreferenceChangeListener((preference, newValue) -> {
//                    VoskInputDevice.deleteCurrentModel(requireContext());
//                    return true;
//                });
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getActivity() instanceof SettingsActivity) {
            // make sure correct title is used even after
            final SettingsActivity settingsActivity = (SettingsActivity) getActivity();
            settingsActivity.toolbarTitle = getString(R.string.pref_io);
            settingsActivity.toolbar.setTitle(R.string.pref_io);
        }
    }
}

File: app/src/main/kotlin/org/stypox/dicio/settings/IOFragment.kt
package org.stypox.dicio.settings

import android.os.Bundle
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import org.stypox.dicio.R
import org.stypox.dicio.io.input.VoskInputDevice

class IOFragment : PreferenceFragmentCompat() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.pref_io)
        findPreference<Preference>(getString(R.string.pref_key_language))
            ?.setOnPreferenceChangeListener { _, _ ->
                //VoskInputDevice.deleteCurrentModel(requireContext())
                activity?.recreate()
                true
            }
        findPreference<Preference>(getString(R.string.pref_key_input_method))
            ?.setOnPreferenceChangeListener { _, _ ->
                VoskInputDevice.deleteCurrentModel(requireContext())
                true
            }
    }

    override fun onResume() {
        super.onResume()
        (activity as? SettingsActivity)?.apply {
            // make sure correct title is used even after
            toolbarTitle = getString(R.string.pref_io)
            toolbar.setTitle(R.string.pref_io)
        }
    }
}
